### PR TITLE
Update of mkFit for 12_1_0_pre2

### PR DESCRIPTION
**Pull request by mmasciov**
#### PR description:

This PR follows #34395 and updates mkFit for physics validation in CMSSW_12_1_0_pre2 (https://its.cern.ch/jira/browse/PDMVRELVALS-130).

In detail, this PR includes:
- update usage of SiStrip quality DB information for mkFit;
- enable usage of SiStrip quality DB information for mkFit;
- enable by default only InitialStepPreSplitting, InitialStep, HighPtTripletStep and DetachedQuadStep in trackingMkFit procModifier;
- other technical optimizations.

It requires cms-sw/cmsdist#7217 and cms-data/RecoTracker-MkFit#3

Developments have been presented at Tracking POG (https://indico.cern.ch/event/1065324/#3-mkfit-status-report) and agreed with PPD (https://indico.cern.ch/event/1063451/#8-mkfit-report).

#### PR validation:

MTV results for tracking-only validation in TTbar RelVal (with mkFit enabled for 4 iterations as specified above):
- on /RelValTTbar_14TeV/CMSSW_12_0_0_pre6-PU_120X_mcRun3_2021_realistic_v4-v1/GEN-SIM-DIGI-RAW: https://mmasciov.web.cern.ch/mmasciov/MkFitCMSSW/MTV_mkFitPR_TTbarPU/
- on /RelValTTbar_14TeV/CMSSW_12_0_0_pre6-PU_120X_mcRun3_2021_realistic_v4_JIRA_129-v1/GEN-SIM-DIGI-RAW: https://mmasciov.web.cern.ch/mmasciov/MkFitCMSSW/MTV_mkFitPR_TTbarPU_JIRA129/

FYI @mtosi @vmariani @mmusich
(@slava77 @makortel)

### CODE REVIEW

**Comment by slava77** [comment]
test parameters:
- workflow = 11634.7
- pull_request = cms-sw/cmsdist#7217,cms-data/RecoTracker-MkFit#3

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc [hunk 1]
```diff
@@ -85,14 +87,52 @@ void MkFitEventOfHitsProducer::produce(edm::StreamID iID, edm::Event& iEvent, co
...
+              << "insert bad range " << first << " to " << last << " " << phi1 << " " << phi2;
+          dv.push_back({phi1, phi2, q1, q2});
+        };
+
+        for (int apv = 0; apv < 6; ++apv) {
+          const bool isBad = bs.BadApvs & (1 << apv);
+          if (isBad)
+            LogTrace("SiStripBadComponents") << "bad apv " << apv << " on " << bs.detid;
+          if (isBad) {
+            if (lastApv == -1) {
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:120]: why not like this? Is there something special about LogTrace?
```suggestion
          if (isBad) {
            LogTrace("SiStripBadComponents") << "bad apv " << apv << " on " << bs.detid;
            if (lastApv == -1) {
```

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc [hunk 2]
```diff
@@ -85,14 +87,52 @@ void MkFitEventOfHitsProducer::produce(edm::StreamID iID, edm::Event& iEvent, co
...
+          float phi1 = firstPoint.phi();
+          float phi2 = lastPoint.phi();
+          if (reco::deltaPhi(phi1, phi2) > 0)
+            std::swap(phi1, phi2);
+          LogTrace("SiStripBadComponents")
+              << "insert bad range " << first << " to " << last << " " << phi1 << " " << phi2;
+          dv.push_back({phi1, phi2, q1, q2});
+        };
+
+        for (int apv = 0; apv < 6; ++apv) {
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:115]: might the apv max number be a named constant?

**Review by jpata — COMMENTED** [review]
The reco changes in the TICL tracksters do not look related to this PR (not sure about the reason at this moment). Minor comments in the code inline.

 @mtosi @vmariani please let me know if you have any comments (with this targeting pre2, closing on Monday)

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc (hunk 1 above)
Comment (slava77) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:120]: yes, the suggestion is OK, I was moving these lines a few too many times.
LogTrace does not have additional header/footer message formatting and is easier to parse than LogDebug.

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc (hunk 2 above)
Comment (slava77) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:115]: @pieterdavid @mmusich 
do we have max APVs and n strips per APV formalized in common constants somewhere?
I tried to git grep, but all that I found was hardcoded 6 and 128.

Comment (mmusich) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:115]: >  do we have max APVs and n strips per APV formalized in common constants somewhere?

they are (at least) in this file: https://github.com/cms-sw/cmssw/blob/master/CalibTracker/SiStripCommon/data/SiStripDetInfo.dat.
Or you can do something like:
```C++
      unsigned short nApvs = stripDet->specificTopology().nstrips() / 128;
```
128 is univeral for all modules, I am not sure there's a constant somewhere.

Comment (pieterdavid) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:115]: the constants are also here: https://github.com/cms-sw/cmssw/blob/master/DataFormats/SiStripCommon/interface/ConstantsForHardwareSystems.h#L38-L43 (6 as max number of APVs indirectly, as 2 per optical link * max 3 of these per module)

Comment (slava77) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc:115]: > what about modules with less than 6 APVs?

it shouldn't matter here, if the `BadApvs` is filled correctly.

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc (hunk 1 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc]: Updated in commit 91d1cafa99953018fe8d4b3223eee72c437cce9c

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc (hunk 2 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc]: Updated in commit 91d1cafa99953018fe8d4b3223eee72c437cce9c

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Thanks for the update. I have no more comments myself. I can sign after waiting until e.g. tomorrow if something comes up from the TRK groups.

**Comment by mmasciov** [comment]
> Thanks for the update. I have no more comments myself. I can sign after waiting until e.g. tomorrow if something comes up from the TRK groups.

As I understand pre2 is closing tomorrow or soon after and there was no new comment since Thursday, is there anything preventing this PR (+related ones) to be merged? Thanks.

**Comment by jpata** [comment]
+reconstruction
- no reco changes in standard workflows (mkFit workflow is not tracked)
- requires cms-sw/cmsdist#7217 and cms-data/RecoTracker-MkFit#3

**Comment by perrotta** [comment]
+operations

**Comment by perrotta** [comment]
@smuzaffar should we wait for you merging cms-sw/cmsdist#7229, or will you do it as soon as we merge this PR in cmsssw?

**Comment by smuzaffar** [comment]
@perrotta , you can also merge the cmsdist PRs ( cms-sw/cmsdist#7217 and  cms-sw/cmsdist#7229 ) when you merge this one. just issue `merge` comment on those PRs :-)

**Comment by perrotta** [comment]
> @perrotta , you can also merge the cmsdist PRs ( [cms-sw/cmsdist#7217](https://github.com/cms-sw/cmsdist/pull/7217) and [cms-sw/cmsdist#7229](https://github.com/cms-sw/cmsdist/pull/7229) ) when you merge this one. just issue `merge` comment on those PRs :-)

Thank you @smuzaffar, I will.
Since I'm kind of new in the game, I just wanted to be sure that I did not bypass your signature without your agreement

**Comment by perrotta** [comment]
+1

**Pull request merged**